Add mocha tests linting

roman-mazur opened the pull request:
As suggested by @Page- 
https://www.flowdock.com/app/rulemotion/resin-tech/messages/5050379

Change-type: minor
Signed-off-by: Roman Mazur <roman@balena.io>

Examples:
```
$ resin-lint --tests ./test/
  ✓ test/00-init.coffee
  ✓ test/01-constants.spec.coffee
  ✓ test/02-db.spec.coffee
  ✓ test/03-config.spec.coffee

✓ Ok! » 0 errors and 0 warnings in 25 files

Mocha tests check FAILED!
File test/01-constants.spec.coffee, line 4: found describe.only

$ resin-lint --typescript --tests ./test/


Mocha tests check FAILED!
File test/local-mode.ts, line 301: found describe.only
File test/local-mode.ts, line 312: found it.only
```

I can check eslint plugin alternatives later - for now this change can help with keeping checks consistent across the repos.

CameronDiver commented on lib/mocha-tests-lint.ts:
```diff
@@ -0,0 +1,34 @@
+	let message = '';
+
+	for (const scriptPath of scripts) {
+		const content = fs.readFileSync(scriptPath, { encoding: 'utf8' });
```
I'd rather this wasn't a synchronous file access.

roman-mazur replied:
Would you elaborate on that? What potentials problems are thinking about?

Page- replied:
In general it blocks the event loop which means nothing else can happen at the same time, and in this case that'll mean it takes longer to lint compared to if the file reads were done in parallel (potentially significantly longer with lots of mocha files)

roman-mazur replied:
Parallelized.

Page- commented on package.json:
```diff
@@ -31,7 +31,7 @@
     "@types/depcheck": "^0.6.0",
     "@types/glob": "^5.0.35",
-    "@types/node": "^8.10.45",
+    "@types/node": "^10.14.10",
```
Why was this necessary?  Mainly asking because it's potentially a breaking change

roman-mazur replied:
should be possible to revert - I needed this because the original implementation was using `fs.readDir` variation that was returning `Dirent[]`, and it was introduced in later versions.

But since I rewrote the checker to fit the rest of the lint code, it's not longer necessary.

Page- commented on lib/mocha-tests-lint.ts:
```diff
@@ -0,0 +1,34 @@
+	) {}
+}
+
+const checkPattern = /((describe)|(it))\.only/;
```
```suggestion
const checkPattern = /(describe|it)\.only/;
```

roman-mazur commented:
@Page- @CameronDiver 
PTAL

roman-mazur commented:
@balena-ci retest

roman-mazur commented:
My concern is also that it will not be enough to bump the linter version to get mocha tests check running - linter needs to understand if running the checks makes sense (--tests cmd line flag).

Do you guys think we should implicitly set this flag if the input directory is `"test"`?

Page- commented on lib/mocha-tests-lint.ts:
```diff
@@ -0,0 +1,50 @@
+	let errorsFound = false;
+	let message = '';
+
+	const readFile = Bluebird.Promise.promisify(fs.readFile);
```
```suggestion
	const readFile = Bluebird.promisify(fs.readFile);
```

Page- commented on lib/mocha-tests-lint.ts:
```diff
@@ -0,0 +1,50 @@
+	for (const scriptPath of scripts) {
+		allChecks.push(
+			readFile(scriptPath)
+				.then(buffer => buffer.toString())
```
```suggestion
				readFile(scriptPath, 'utf8')
```

Page- commented on lib/mocha-tests-lint.ts:
```diff
@@ -0,0 +1,50 @@
+	const readFile = Bluebird.Promise.promisify(fs.readFile);
+
+	const allChecks = [];
+	for (const scriptPath of scripts) {
```
You can do this as
```ts
const allChecks = scripts.map(async (script) => {
	content = await readFile(scriptPath, 'utf8')
	...
})

roman-mazur replied:
Speaking about `readFile(scriptPath, 'utf8')`, it does not compile. Type of `Bluebird.promisify(fs.readFile)` allows only one argument.

Page- replied:
Yeah, that's a typing issue because if you did
```ts
const readFile = Bluebird.Promise.promisify(fs.readFile) as any;
```
then it would work fine at runtime.  I normally use `fs.promises.readFile` in node 10+ to get around this because it has the built in types that work nicely but we can't guarantee that for resin-lint :(

Page- commented on lib/resin-lint.ts:
```diff
@@ -1,4 +1,4 @@
-import * as Promise from 'bluebird';
+import * as BluePromise from 'bluebird';
```
```suggestion
import * as Bluebird from 'bluebird';
```

CameronDiver replied:
We generally do `import * as Bluebird from 'bluebird';` and you've also done this in mocha-test-lint.ts, so it'd be better to stay consistent.

CameronDiver commented on lib/mocha-tests-lint.ts:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	try {
+		await Promise.all(allChecks);
```
`await Bluebird.map` with an async callback would make more sense here.

roman-mazur replied:
I've used Page's suggestion above.

roman-mazur commented:
@balena-ci retest

I fixed Bluebird imports and removed `for` loop in the mocha checks implementation.

What do you think now, @Page- @CameronDiver ?

roman-mazur commented:
Also, any thoughts on doing tests check using the source directory name?
https://github.com/balena-io/node-resin-lint/pull/39#issuecomment-508122008

CameronDiver commented:
My concern is that I'd only want this to run on commit/prepare. I'm fine with `test(s?)` being used as the directory we search, but I'd like a flag either enabling or disabling this checking. I'm not too familiar with optimist but it looks like you have a flag for these checks, is it automatically enabled? How would I then disable it?

roman-mazur commented:
@CameronDiver I illustrated usage in supervisor (as you wanted, as a pre-commit hook only) here:
https://github.com/balena-io/balena-supervisor/pull/1031

Does it answer your question?
